File: code/ex886.java
class Solution {
    public boolean possibleBipartition(int N, int[][] dislikes) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int[] dislike : dislikes) {
            List<Integer> d1 = map.getOrDefault(dislike[0], new ArrayList<>());
            d1.add(dislike[1]);
            map.put(dislike[0], d1);
            List<Integer> d2 = map.getOrDefault(dislike[1], new ArrayList<>());
            d2.add(dislike[0]);
            map.put(dislike[1], d2);
        }
        
        int[] root = new int[N + 1];
        for (int i = 0; i <= N; i++)    root[i] = i;
        for (Map.Entry<Integer, List<Integer>> entry : map.entrySet()) {
            int r = find(root, entry.getKey());
            List<Integer> list = entry.getValue();
            for (int i = 0; i < list.size(); i++) {
                if (find(root, list.get(i)) == r)    return false;
                for (int j = i + 1; j < list.size(); j++) {
                    union(root, list.get(i), list.get(j));
                }
            }
        }
        
        return true;
    }
    
    private void union(int[] root, int p1, int p2) {
        int r1 = find(root, p1), r2 = find(root, p2);
        if (r1 <= r2) {
            root[r2] = r1;
        } else {
            root[r1] = r2;
        }
    }
    
    private int find(int[] root, int p) {
        if (root[p] != p)
            root[p] = find(root, root[p]);
        return root[p];
    }
}

File: code/ex886.cpp
class Solution {
public:
    bool possibleBipartition(int N, vector<vector<int>>& dislikes) {
        vector<vector<int>> dislike(N + 1, vector<int>());
        for (auto d : dislikes) {
            dislike[d[0]].push_back(d[1]);
            dislike[d[1]].push_back(d[0]);
        }
        
        vector<int> root(N + 1);
        for (int i = 1; i <= N; i++)    root[i] = i;
        for (int d = 1; d <= N; d++) {
            int rd = find(root, d);
            if (dislike[d].size() > 0) {
                int r0 = find(root, dislike[d][0]);
                for (int i = 1; i < dislike[d].size(); i++) {
                    int ri = find(root, dislike[d][i]);
                    if (ri == rd)    return 0;
                    uni(root, ri, r0);
                }
            }
        }
        
        return 1;
    }
    
private:
    int find(vector<int>& root, int a) {
        if (root[a] != a)
            root[a] = find(root, root[a]);
        return root[a];
    }
    
    void uni(vector<int>& root, int a, int b) {
        int ra = find(root, a), rb = find(root, b);
        if (ra < rb)
            root[ra] = rb;
        else
            root[rb] = ra;
    }
};